### PR TITLE
bpo-34963: Make the repr of the typing.NewType() result more meaningful.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Add attributes `__qualname__` and `__module__`.
If the name argument is dotted, the `__name__` attribute will be set to its last component.

https://bugs.python.org/issue34963

### CODE REVIEW

**Review by ilevkivskyi — COMMENTED** [review]
I actually like this solution. But before merging let us wait for the benchmarks for other approach, if they are OK, maybe we can have an even nicer repr in https://github.com/python/cpython/pull/9808

File: Lib/typing.py
Comment (vstinner) [inline Lib/typing.py]: It looks like a line is duplicated?

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/typing.py
Comment (serhiy-storchaka) [reply Lib/typing.py]: Thanks, removed.

**Review by vstinner — APPROVED** [review]
LGTM, but I would prefer that one of the typing maintainer double check this change. @ilevkivskyi?

**Comment by ilevkivskyi** [comment]
@vstinner The PR looks good. However, there is a "competing" PR (alternative solution) in #9808. I asked the author for benchmarks, but he didn't respond yet. If he will not respond next week, I would propose to merge this one.

**Comment by vstinner** [comment]
@serhiy-storchaka: This PR still LGTM. You can merge it.

> However, there is a "competing" PR (alternative solution) in #9808.

It doesn't seem to define __module__. It doesn't seem to be directly related. This PR looks simple enough.

**Comment by ilevkivskyi** [comment]
It looks like the author of other PR is not responding. I leave up to Serhiy whether to just move with this PR or incorporate some ideas from other one too as I proposed in https://bugs.python.org/issue34963#msg328707

(This PR LGTM as is.)

**Comment by gaborbernat** [comment]
@serhiy-storchaka shall we move ahead with this or the alternate PR?

**Comment by JelleZijlstra** [comment]
This PR still applies and it would be useful, is there anything blocking it from being merged?

**Review by gvanrossum — APPROVED** [review]
Okay, looks fine to me. I'll merge.

**Comment by gvanrossum** [comment]
Closing and reopening since it looks like some tests are hanging.

**Comment by gvanrossum** [comment]
@serhiy-storchaka Can you look at the test failure? Maybe it just needs a merge from a more recent main branch?

**Comment by serhiy-storchaka** [comment]
It needs some correction in `is_new_type()` in `Objects/unionobject.c`.

**Comment by gvanrossum** [comment]
Go for it!

**Comment by serhiy-storchaka** [comment]
typing.NewType has been reimplemented as a class in bpo-44353, so this trick is no longer needed.

**Pull request closed without merging**
